### DISCUSSION TITLE
Does VCR provide an easy way to discard episodes which lead to a test failure?

### DISCUSSION THREAD

**Discussion by Adamantish** [vcr/vcr#854] [Q&A]
I guess I'm talking about a possible helper method setup to detect a test failure in some of the common frameworks like rspec, cucumber, etc...

For some context, what we're trying to do is come up with a configuration that makes VCR a set and forget kind of thing. Something that just applies to our whole test suite.

We're most of the way there. It generates cassette names based on the test example name and some other helper mechanisms.
But we have some tests that already include a retrying mechanism which we need to keep but we want to discard interactions that result in a test failure.

The hack we're currently having to do is recording the cassette state before each example and reverting it in memory - `VCR.current_cassette` - and the yaml file.

**Comment by mcfiredrill** [comment]
Hmmm, I thought cassettes weren't recorded on test failure...

**Comment by Adamantish** [comment]
Ok, well the trouble here is that test failures are being suppressed so as to do the retry. If the test eventually passes on a subsequent retry then all the failing calls leading up to there get recorded in the cassette before the final good one. Am I right in thinking that without receiving a test pass or failure message, that VCR treats these as just a lot of identical, undemarcated assertions within one test example? That it builds up a list of all the interactions before finally hitting the `before_record` hook upon test pass/failure? Or just pass?

**Comment by Adamantish** [comment]
@myronmarston Can you confirm what @mcfiredrill says. 
They're certainly being recorded for us and I'm trying to get the bottom of why.

**Comment by myronmarston** [comment]
VCR knows nothing about test success or failure.  It doesn't generally even know anything about test frameworks (notice that virtually all of the relish examples are written as stand-alone scripts).  There is support for cucumber and RSpec, of course, but that's just provides convenient integration, and doesn't change the fact that VCR's core logic knows nothing about tests.

**Comment by Adamantish** [comment]
Thanks. Sounds like it could do with an rspec-vcr gem for deeper integration. In principle it should be possible to just apply VCR to the whole of an existing test suite and then not worry about it. We have made most of that extra framework (just auto-naming folders and cassette files after the spec file names and example descriptions respectively). We now have the difficulty of having to constantly find cassettes which correspond to failed tests so we can delete them.

**Comment by galori** [comment]
Is this for CI or for Dev? 

Generally tests that are not changing and are stable need not get rerecorded, you should already have a cassette for them. 

You can set the default to not record new cassettes (or record once) and only when a developer is adding or modifying a test suite they set it to record just to generate the cassettes for that new test suite. Before the new test is committed it's set back to the default.

How does your Dev process work?

**Comment by selimerunkut** [comment]
> Thanks. Sounds like it could do with an rspec-vcr gem for deeper integration. In principle it should be possible to just apply VCR to the whole of an existing test suite and then not worry about it. We have made most of that extra framework (just auto-naming folders and cassette files after the spec file names and example descriptions respectively). We now have the difficulty of having to constantly find cassettes which correspond to failed tests so we can delete them.

A little late to to the party. Had the exact same problem, especially for selenium and integration test;  how about this solution: 

rails_helper.rb:

```
VCR.configure do |c|
  vcr_mode = ENV['VCR_MODE'] =~ /rec/i ? :all : :once

  c.default_cassette_options = {
    record: vcr_mode,
    match_requests_on: %i[method uri body]
  }
end
```
how to use:

`VCR_MODE=rec bundle exec rspec spec/some_class_spec.rb:30
`
credits go to: 
https://fabioperrella.github.io/10_tips_to_help_using_the_VCR_gem_in_your_ruby_test_suite.html
